spark/dl/src/main/scala/com/intel/analytics/bigdl/dataset/DataSet.scala

```diff
@@ -357,13 +443,65 @@ object DataSet {
    */
   def rdd[T: ClassTag](data: RDD[T]): DistributedDataSet[T] = {
     val nodeNumber = Engine.nodeNumber()
-    new CachedDistriDataSet[T](
-      data.coalesce(nodeNumber, true)
-        .mapPartitions(iter => {
-          Iterator.single(iter.toArray)
-        }).setName("cached dataset")
-        .cache()
-    )
+    val transformedData = if (declineRepartitionedRdd) {
+      new CachedDistriDataSet[T](
+        data.coalesce(nodeNumber, true)
+          .mapPartitions(iter => {
+            Iterator.single(iter.toArray)
+          }).setName("cached dataset")
+          .cache()
+      )
+    } else {
+      val originalRdd = data
+
+      // should not call methods other than transform() and getSplits(), for that will be meaningless
+      new DistributedDataSet[T] with Serializable {
+        override def originRDD(): RDD[_] = originalRdd
+        override def data(train: Boolean): RDD[T] = ???
+        override def shuffle(): Unit = ???
+        override def size(): Long = ???
+
+        override def transform[C: ClassTag](transformer: Transformer[T, C]): DataSet[C] = {
+          new SplittableTransformedDistributedDataSet(this, transformer)
+        }
+
+        override def getSplits(): Seq[CachedDistriDataSet[T]] = {
+          if (declineRepartitionedRdd) {
+            Seq(new CachedDistriDataSet[T](
+              originalRdd.coalesce(nodeNumber, true)
+                .mapPartitions(iter => {
+                  Iterator.single(iter.toArray)
+                }).setName("cached dataset")
+                .cache()
+            ))
+          } else {
+            val nodeNum = Engine.nodeNumber()
+            val partitions = originalRdd.partitions.length
+            val splits = partitions / nodeNum + (if (partitions % nodeNum == 0) 0 else 1)
+
+            val rddBatches = mutable.ArrayBuffer[CachedDistriDataSet[T]]()
+            for (i <- 0 until splits) {
+              val splittedData = originalRdd
+                .mapPartitionsWithIndex((index, iter) => {
+                  val upperLimit = (i + 1) * nodeNum
+                  val downLimit = upperLimit - nodeNum
+                  if (index < upperLimit && index >= downLimit) iter
+                  else Iterator.empty
+                })
+              rddBatches.append(new CachedDistriDataSet(
```

Comment:
@songhappy I guess this solution is similar with what you did which split a "big" RDD into smaller pieces and then cache it in turn. One concern is we need to go through the whole dataset to filter the data each time when switching to a new "smaller" RDD. It might not be a issue as it only happen after several epochs and also depends on how much time it would take.

Reply:
@zhichao-li yes it will trigger rescanning the original data. When we have enough executor, it only takes a short time. But more executor means more model and longer all-reduce procedure, thus  should be deliberately considered. (btw, we can easily cache all train data（100GB） in memory to accelerate the procedure, as it's much smaller than `Sample` or `Tensor`)